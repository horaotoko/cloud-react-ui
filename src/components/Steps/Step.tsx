import { forwardRef } from "react";
import styled, {css} from "styled-components";
import { defaultProps } from "../../theme";

export type Props = {
  step: number;
  stepStatus: 'waiting' | 'inProgress' | 'completed';
  title: string;
  useCheck: boolean;
  className?: string;
}

const StepWrapper = styled.div`
  ${( {theme: { steps } } ) => css`
    position: relative;
  `}
`;
StepWrapper.defaultProps = defaultProps

const Title = styled.span<{ stepStatus: Props['stepStatus'] }>`
  ${( {theme: { steps }, stepStatus } ) => css`
    color: ${steps.step.title.color[stepStatus]};
    position: ${steps.step.title.position};
    top: ${steps.step.title.top};
    left: ${steps.step.title.left};
    transform: ${steps.step.title.transform};
    min-width: ${steps.step.title.minWidth};
    text-align: ${steps.step.title.textAlign};
    font-size: ${steps.step.title.fontSize};
    font-weight: ${steps.step.title.fontWeight};
  `}
`;
Title.defaultProps = defaultProps

const Circle = styled.div<{ stepStatus: Props['stepStatus'] }>`
  ${( {theme: { steps }, stepStatus } ) => {
    return css`
      margin-top: ${steps.step.marginTop};
      width: ${steps.step.width};
      height: ${steps.step.height};
      border-radius: ${steps.step.borderRadius};
      display: ${steps.step.display};
      align-items: ${steps.step.alignItems};
      justify-content: ${steps.step.justifyContent};
      background-color: ${steps.step.backgroundColor[stepStatus]};
      border-style: ${steps.step.borderStyle};
      border-width: ${steps.step.borderWidth};
      border-color: ${steps.step.borderColor[stepStatus]};
      color: ${steps.step.color[stepStatus]};
      line-height: ${steps.step.lineHeight};
    `
  }}
`;
Circle.defaultProps = defaultProps

const Step = forwardRef<HTMLDivElement, Props>(
  ({step, title, useCheck,  className, ...rest}, ref) => (
    <StepWrapper className={className}>
      <Title {...rest}>
        {title}
      </Title>
      { useCheck }
      <Circle ref={ref} {...rest}>
        {step}
      </Circle>
    </StepWrapper>
  )
);

Step.displayName = "Step"

export { Step };
