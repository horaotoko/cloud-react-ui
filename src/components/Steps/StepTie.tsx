import { forwardRef } from "react";
import styled, {css} from "styled-components";
import { defaultProps } from "../../theme";

export type Props = {
  completed?: boolean;
  className?: string;
}

const StyledStepTie = styled.div<Props>`
  ${({ theme: { steps }, completed = false }) => css`
    width: ${steps.stepTie.width};
    height: ${steps.stepTie.height};
    background: ${completed ? steps.stepTie.background["completed"] : steps.stepTie.background["waiting"]};
  `}
`;
StyledStepTie.defaultProps = defaultProps

const StepTie = forwardRef<HTMLDivElement, Props>((props, ref) => (
  <StyledStepTie ref={ref} {...props} />
));

StepTie.displayName = "StepTie"

export { StepTie };
