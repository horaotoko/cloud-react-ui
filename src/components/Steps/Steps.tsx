import { createContext, forwardRef } from 'react';
import styled, { css } from 'styled-components';
import { defaultProps } from '../../theme';

export type Mode = 'number' | 'check';
export const StepsModeContext = createContext<{ mode: Mode }>({
  mode: 'check',
});

export type Props = {
  className?: string;
  mode?: Mode;
  children: React.ReactNode;
};

const StyledSteps = styled.div<Props>`
  ${({ theme: { steps } }) => css`
    display: ${steps.display};
    align-items: ${steps.alignItems};
  `}
`;
StyledSteps.defaultProps = defaultProps;

const Steps = forwardRef<HTMLDivElement, Props>(
  ({ mode = 'check', ...rest }, ref) => (
    <StepsModeContext.Provider value={{ mode }}>
      <StyledSteps ref={ref} {...rest} />
    </StepsModeContext.Provider>
  )
);

Steps.displayName = 'Steps';

export { Steps };
