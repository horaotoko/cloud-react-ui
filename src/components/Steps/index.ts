export { Steps } from './Steps';
export type { Props as StepsProps } from './Steps';

export { Step } from './Step';
export type { Props as StepProps } from './Step';

export { StepTie } from './StepTie';
export type { Props as StepTieProps } from './StepTie';
