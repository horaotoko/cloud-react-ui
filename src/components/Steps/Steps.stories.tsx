import { Story, Meta } from '@storybook/react';
import { Steps } from "./Steps";
import { StepTie } from "./StepTie"

export default {
  component: Steps,
  title: "components/Steps"
} as Meta;

const Template: Story = () => {
  return (
    <Steps>
      <StepTie completed />
      <StepTie />
    </Steps>
  );
};

export const Sample = Template.bind({});
