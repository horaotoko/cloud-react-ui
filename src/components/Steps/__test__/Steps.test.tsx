import { Steps } from "../Steps";
import { Step } from "../Step";
import { StepTie } from "../StepTie";
import { render, screen } from "@testing-library/react";

describe("ui/Steps", () => {
  it("renders of check mode", () => {
    const { asFragment } = render(
      <Steps>
        <Step step={1} stepStatus="completed" title="Step 1" />
        <StepTie completed />
        <Step step={2} stepStatus="inProgress" title="Step 2" />
        <StepTie />
        <Step step={3} stepStatus="waiting" title="Step 3" />
      </Steps>
    );
    expect(screen.getByTitle('check')).toBeInTheDocument();
    expect(screen.queryByText('1')).not.toBeInTheDocument();
    expect(screen.getByText('2')).toBeInTheDocument();
    expect(screen.getByText('3')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders of number mode", () => {
    const { asFragment } = render(
      <Steps mode="number">
        <Step step={1} stepStatus="completed" title="Step 1" />
        <StepTie completed />
        <Step step={2} stepStatus="inProgress" title="Step 2" />
        <StepTie />
        <Step step={3} stepStatus="waiting" title="Step 3" />
      </Steps>
    );
    expect(screen.getByText('1')).toBeInTheDocument();
    expect(screen.getByText('2')).toBeInTheDocument();
    expect(screen.getByText('3')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
