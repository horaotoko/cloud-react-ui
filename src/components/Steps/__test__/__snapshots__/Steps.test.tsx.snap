// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/Steps renders of check mode 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c1 {
  position: relative;
}

.c2 {
  color: #3B7DE9;
  position: absolute;
  top: 0%;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  min-width: 110px;
  text-align: center;
  font-size: 13px;
  font-weight: bold;
}

.c7 {
  color: #666;
  position: absolute;
  top: 0%;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  min-width: 110px;
  text-align: center;
  font-size: 13px;
  font-weight: bold;
}

.c3 {
  margin-top: 26px;
  width: 34px;
  height: 34px;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #3B7DE9;
  border-style: solid;
  border-width: 1px;
  border-color: #3B7DE9;
  color: #FFFFFF;
  line-height: 13px;
}

.c8 {
  margin-top: 26px;
  width: 34px;
  height: 34px;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #FFFFFF;
  border-style: solid;
  border-width: 1px;
  border-color: #3B7DE9;
  color: #3B7DE9;
  line-height: 13px;
}

.c11 {
  margin-top: 26px;
  width: 34px;
  height: 34px;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #D4D8DD;
  border-style: solid;
  border-width: 1px;
  border-color: #D4D8DD;
  color: #FFFFFF;
  line-height: 13px;
}

.c9 {
  position: relative;
  top: -1px;
}

.c4 {
  width: 18px;
  height: 14px;
}

.c4 > path {
  fill: #FFFFFF;
}

.c5 {
  width: 110px;
  height: 36px;
  margin: 0 6px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  width: 100%;
  height: 2px;
  background: #3B7DE9;
}

.c10 {
  width: 100%;
  height: 2px;
  background: #D4D8DD;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        Step 1
      </span>
      <div
        class="c3"
      >
        <svg
          class="c4"
          fill="none"
          viewBox="0 0 18 14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            check
          </title>
          <path
            clip-rule="evenodd"
            d="M5.9999 11.2L1.7999 6.99998L0.399902 8.39998L5.9999 14L17.9999 1.99998L16.5999 0.599976L5.9999 11.2Z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
    </div>
    <div
      class="c5"
    >
      <div
        class="c6"
      />
    </div>
    <div
      class="c1"
    >
      <span
        class="c7"
      >
        Step 2
      </span>
      <div
        class="c8"
      >
        <span
          class="c9"
        >
          2
        </span>
      </div>
    </div>
    <div
      class="c5"
    >
      <div
        class="c10"
      />
    </div>
    <div
      class="c1"
    >
      <span
        class="c7"
      >
        Step 3
      </span>
      <div
        class="c11"
      >
        <span
          class="c9"
        >
          3
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ui/Steps renders of number mode 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c1 {
  position: relative;
}

.c2 {
  color: #3B7DE9;
  position: absolute;
  top: 0%;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  min-width: 110px;
  text-align: center;
  font-size: 13px;
  font-weight: bold;
}

.c7 {
  color: #666;
  position: absolute;
  top: 0%;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  min-width: 110px;
  text-align: center;
  font-size: 13px;
  font-weight: bold;
}

.c3 {
  margin-top: 26px;
  width: 34px;
  height: 34px;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #3B7DE9;
  border-style: solid;
  border-width: 1px;
  border-color: #3B7DE9;
  color: #FFFFFF;
  line-height: 13px;
}

.c8 {
  margin-top: 26px;
  width: 34px;
  height: 34px;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #FFFFFF;
  border-style: solid;
  border-width: 1px;
  border-color: #3B7DE9;
  color: #3B7DE9;
  line-height: 13px;
}

.c10 {
  margin-top: 26px;
  width: 34px;
  height: 34px;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #D4D8DD;
  border-style: solid;
  border-width: 1px;
  border-color: #D4D8DD;
  color: #FFFFFF;
  line-height: 13px;
}

.c4 {
  position: relative;
  top: -1px;
}

.c5 {
  width: 110px;
  height: 36px;
  margin: 0 6px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  width: 100%;
  height: 2px;
  background: #3B7DE9;
}

.c9 {
  width: 100%;
  height: 2px;
  background: #D4D8DD;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        Step 1
      </span>
      <div
        class="c3"
      >
        <span
          class="c4"
        >
          1
        </span>
      </div>
    </div>
    <div
      class="c5"
    >
      <div
        class="c6"
      />
    </div>
    <div
      class="c1"
    >
      <span
        class="c7"
      >
        Step 2
      </span>
      <div
        class="c8"
      >
        <span
          class="c4"
        >
          2
        </span>
      </div>
    </div>
    <div
      class="c5"
    >
      <div
        class="c9"
      />
    </div>
    <div
      class="c1"
    >
      <span
        class="c7"
      >
        Step 3
      </span>
      <div
        class="c10"
      >
        <span
          class="c4"
        >
          3
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
